lib/bats-core/test_functions.bash

```diff
@@ -3,30 +3,83 @@
 BATS_TEST_DIRNAME="${BATS_TEST_FILENAME%/*}"
 BATS_TEST_NAMES=()
 
-# Shorthand for source-ing files relative to the BATS_TEST_DIRNAME,
-# optionally with a .bash suffix appended. If the argument doesn't
-# resolve relative to BATS_TEST_DIRNAME it is sourced as-is.
+# _load expects an absolute path to a file or directory to source.
+# If the argument is a file the file will be sourced or the same path
+# with the added suffix `.bash` exists it will be sourced..
+# If the argument is a directory _load checks of the file `load` or
+# `load.bash` inside - if it exists this file is loaded. If it doesn't
+# exist all files with the `.bash` suffix are sourced.
+#
+# Order with `/path/to/example` as argument:
+#  - /path/to/example.bash
+#  - /path/to/example
+#  - /path/to/example/load.bash
+#  - /path/to/example/load
+#  - /path/to/example/*.bash
+_load() {
+    local file="${1:?}"
+
+    if [[ "${file:0:1}" != "/" ]]; then
+        printf "Received argument with a relative path, expected absolute: %s\n", "$file"
+        return 1
+    fi
+
+    local -a opts=(
+        "$file.bash"
+        "$file"
+        "$file/load.bash"
+        "$file/load"
+    )
+
+    for opt in "${opts[@]}"; do
+        if [[ -f "$opt" ]]; then
+            source "$opt"
+            return
+        fi
+    done
+
+    if [[ -d "$file" ]]; then
+        for contained in "$file"/*.bash; do
+            source "$contained"
+        done
+        return
+    fi
+
+    return 1
+}
+
+# Shorthand to source files relative to the test file
+# (BATS_TEST_DIRNAME) and from BATS_LIB_PATH.
 load() {
   local file="${1:?}"
 
-  # For backwards-compatibility first look for a .bash-suffixed file.
-  # TODO consider flipping the order here; it would be more consistent
-  # and less surprising to look for an exact-match first.
-  if [[ -f "${BATS_TEST_DIRNAME}/${file}.bash" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}.bash"
-  elif [[ -f "${BATS_TEST_DIRNAME}/${file}" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}"
+  # Check if target is absolute
+  if [[ "${file:0:1}" == "/" ]]; then
+    if ! _load "$file"; then
+      printf "Failed to load file '%s'\n" "$file" >&2
+      exit 1
+    fi
+    return
   fi
 
-  if [[ ! -f "$file" ]] && ! type -P "$file" >/dev/null; then
-    printf 'bats: %s does not exist\n' "$file" >&2
-    exit 1
+  local bats_lib_path="$BATS_LIB_PATH"
+  if [[ -z "$bats_lib_path" ]]; then
+    bats_lib_path="$HOME/.bats/lib:/usr/lib/bats"
   fi
+  bats_lib_path="$BATS_TEST_DIRNAME:$bats_lib_path"
+
+  local -a parts
+  IFS=: read -ra parts <<< "$bats_lib_path"
+
+  for part in "${parts[@]}"; do
+    if _load "$part/$file"; then
+      # _load finished without error, file/library was sourced, return
```

Comment:
I prefer it shorter:

> First successful file/library load wins.

Reply:
Only the first found wins - `_load` does not go through the entire path and _tries_ any targets. The first match it finds it sources and then returns.

Reply:
Yeah, with the changes you suggested to use `return` instead of `exit` the first successfull library wins. Imho that would be wrong though - the first matching library should be sourced and if that fails an error should be reported. I'll update that accordingly.